streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -121,11 +163,13 @@ public void onPartitionsAssigned(Collection<TopicPartition> assignment) {
             try {
                 log.info("stream-thread [{}] New partitions [{}] assigned at the end of consumer rebalance.",
                         StreamThread.this.getName(), assignment);
+                state = StreamThreadStateType.PARTITIONS_ASSIGNED;
                 addStreamTasks(assignment);
                 addStandbyTasks();
                 lastCleanMs = time.milliseconds(); // start the cleaning cycle
                 streamsMetadataState.onChange(partitionAssignor.getPartitionsByHostState(), partitionAssignor.clusterMetadata());
                 initialized.set(true);
```

Comment:
sure. but you could just make a `setState` like you did for `KafkaStreams` and make that sync. 

in fact, right now the `KafkaStreams.StateListener` is not really correct, it only fires `onChange` for a subset of the state transitions. if we make this class follow the same listener pattern  then `KafkaStreams.StateListener.onChange` can fire whenever one of it's `StreamThread`s changes state.